src/PixelFlags.cc

```diff
@@ -57,87 +59,153 @@ class FootprintBits : public afw::detection::FootprintFunctor<MaskedImageT> {
+                                      afw::table::SourceRecord & measRecord) const{
+        lsst::afw::image::MaskPixel bitmask = 0x0;
+        for (KeyMap::const_iterator i = maskFlagToPixelFlag->begin(); i != maskFlagToPixelFlag->end(); ++i){
+            try {
+                bitmask = MaskedImageT::Mask::getPlaneBitMask(i->first);
+            } catch (pex::exceptions::InvalidParameterError & err){
+                throw LSST_EXCEPT(FatalAlgorithmError, err.what());
+            }
+            if (func->getBits() & MaskedImageT::Mask::getPlaneBitMask(i->first)){
+                measRecord.set(i->second, true);
+            }
+        }
+    }
 
 void PixelFlagsAlgorithm::measure(
     afw::table::SourceRecord & measRecord,
     afw::image::Exposure<float> const & exposure
 ) const {
 
-    afw::geom::Point2D center = _centroidExtractor(measRecord, _flagHandler);
-    typedef afw::image::MaskedImage<float> MaskedImageT;
     MaskedImageT mimage = exposure.getMaskedImage();
-
     FootprintBits<MaskedImageT> func(mimage);
-
-//  Catch NAN in centroid estimate
-    if (lsst::utils::isnan(center.getX()) || lsst::utils::isnan(center.getY())) {
-        throw LSST_EXCEPT(pex::exceptions::InvalidParameterError,
-                          "Center point passed to PixelFlagsALgorithm is NaN");
+    
+    // Check if the measRecord has a valid centroid key, i.e. it was
+    // centroided
+    afw::geom::Point2D center;
+    if (measRecord.getTable()->getCentroidKey().isValid()) {
+        center = measRecord.getCentroid(); 
+        //  Catch NAN in centroid estimate
+        if (lsst::utils::isnan(center.getX()) || lsst::utils::isnan(center.getY())) {
+            throw LSST_EXCEPT(pex::exceptions::RuntimeError,"Center point passed to PixelFlagsAlgorithm is Nan");
+        }
+    }
+    else {
+        // Attempting to set pixel flags with no centroider, the center will
+        // be determined though the peak pixel
+        PTR(afw::detection::Footprint) footprint = measRecord.getFootprint();
+        // If there is no footprint or the footprint contains no peaks, throw
+        // a runtime error.
+        if (!footprint || footprint->getPeaks().empty()){
+            throw LSST_EXCEPT(pex::exceptions::RuntimeError,"No footprint, or no footprint peaks detected");
+        }
+        else {
+            center.setX(footprint->getPeaks().front().getFx());
+            center.setY(footprint->getPeaks().front().getFy());
+        }
     }
-//  Catch centroids off the image
+
+    //  Initialize bitmask to use in try statments
+    lsst::afw::image::MaskPixel bitmask = 0x0;
+
+    //  Catch centroids off the image
     if (!mimage.getBBox().contains(afw::geom::Point2I(center))) {
-       _flagHandler.setValue(measRecord, EDGE, true);
+        measRecord.set(_offImageKey, true);
+        auto iter = _predefinedAnyKeys.find("EDGE");
+        assert(iter != _predefinedAnyKeys.end());
+        measRecord.set(iter->second, true);
```

Comment:
The last three lines could be shortened to just `measRecord.set(_predefinedAnyKeys.at("EDGE"), true);`.